fake_tensor.py

```diff
@@ -0,0 +1,162 @@
+import torch
+
+from base_tensor import BaseTensor
+from torch.testing._internal.common_utils import (
+    run_tests,
+    TestCase,
+)
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.fx.operator_schemas import normalize_function
+from typing import Union
+import unittest
+
+aten = torch.ops.aten
+
+
+# Meta tensors give you the ability to run PyTorch code without having to
+# actually do computation through tensors allocated on a `meta` device.
+# Because the device is `meta`, meta tensors do not model device propagation.
+# FakeTensor extends MetaTensors to also carry an additional `fake_device`
+# which tracks devices that would have been used.
+
+
+class FakeTensor(BaseTensor):
+    fake_device: torch.device
+
+    @staticmethod
+    def __new__(cls, elem, device):
+        return super().__new__(cls, elem)
+
+    def __init__(self, elem, device: Union[torch.device, str]):
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        meta_device = torch.device(type="meta")
+        assert elem.device == meta_device
+        assert device != meta_device
+        self.fake_device = (
+            device if isinstance(device, torch.device) else torch.device(device)
+        )
+
+    @staticmethod
+    def from_tensor(t):
+        existing_device = t.device
+        return FakeTensor(t.to(device="meta"), existing_device)
+
+    @property
+    def device(self):
+        return self.fake_device
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        # Run the original computation
+        r = super().__torch_dispatch__(func, types, args, kwargs)
+        kwargs = kwargs if kwargs else {}
+
+        def wrap(e, device):
+            # inplace ops can return fake tensors
+            if isinstance(e, torch.Tensor) and not isinstance(e, cls):
+                return FakeTensor(e, device)
+            else:
+                return e
+
+        # Pytorch device is kwarg-only, except for _pin_memory/pin_memory
+        # which is not yet supported by meta-tensors
+        assert (
+            func != aten._pin_memory.default and func != aten.pin_memory.default
+        ), f"NYI: {func}"
+
+        # if device is specified, use that
+        # not sure this is actually needed.. device only shows up in constructors
+        if kwargs.get("device", None):
+            return tree_map(partial(wrap, device=kwargs["device"]), r)
```

Comment:
This rule is not "sound", in the sense that someone could add an operator with device kwarg used in a non-standard way which would break the calculation here. We can keep doing this cheap and cheerful calculation though if we added a tag; we'd just go through and tag all the device operators with a tag "yes the output tensor matches what device you pass in here" LOL. Is this the right ontology? Probably not. But it's easy and gives a bit more safety.